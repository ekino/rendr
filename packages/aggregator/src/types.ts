import { BlockDefinition, Page, RendrCtx } from "@ekino/rendr-core";

export type HandlerList = {
  [index: string]: Handler;
};

export type Handler = (
  definition: BlockDefinition,
  ctx: RendrCtx,
  page: Page
) => Promise<BlockDefinition>;

export type HandlerRegistry = (code: string) => Handler;
