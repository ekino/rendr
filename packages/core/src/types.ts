import { IncomingMessage, ServerResponse } from "http";

export interface BlockDefinition {
  type: string;
  settings: Settings;
  container: string;
  order: number;
}
export interface Settings {
  [index: string]: any;
}

export interface Map {
  [index: string]: string;
}

export interface Cache {
  ttl: number;
}

export interface Head {
  titleTemplate: string;
  defaultTitle: string;
  title: string;
  link: string;
  htmlAttributes: Settings;
  meta: Settings[];
}

// -- Page definition
export class Page {
  public statusCode: number = 200;
  public type: string = "document";
  public template: string = "default";
  public cache: Cache = {
    ttl: 0
  };
  public head: Head = {
    titleTemplate: "%s",
    defaultTitle: "-",
    title: "-",
    link: "",
    htmlAttributes: {},
    meta: []
  };
  public blocks: BlockDefinition[] = [];
  public settings: Settings = {};
  public id: string = "";
  public path: string = "";
}

// -- NextJs Controller signature
export interface RequestCtx {
  hostname: string; // the hostname requested
  pathname: string; // the path - without the query string
  query: Map; // the query string, parsed
  params: Map; // the params from the routing, ie param from nice url
  asPath: string; // the complete path - pathname + query
  isServerSide: boolean;
  isClientSide: boolean;
  req: IncomingMessage;
  res: ServerResponse;
}

export interface Normalizer {
  (entry: any): BlockDefinition;
}

export type NormalizerList = {
  [index: string]: Normalizer;
};
