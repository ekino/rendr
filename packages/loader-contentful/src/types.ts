import {
  Asset as ContentfulAsset,
  Entry,
  ContentfulCollection
} from "contentful";
import { RequestCtx, BlockDefinition } from "@ekino/rendr-core";

export { Asset as ContentfulAsset } from "contentful";

import { ContentfulClientApi } from "contentful";

export interface EntryNormalizer {
  (entry: Entry<any> | ContentfulAsset, normalizers?: EntryNormalizer):
    | any
    | void;
}

export type EntryNormalizerList = {
  [index: string]: EntryNormalizer;
};

export interface ClientFactory {
  (ctx: RequestCtx): ContentfulClientApi;
}

export interface Website {
  id: string;
  name: string;
  domains: string[];
  path: string;
  culture: string;
  countryCode: string;
  order: number;
  enabled: boolean;
}

export interface Author {
  id: string;
  name: string;
  jobTitle: string;
  slug: string;
  social: {
    twitter: string;
    facebook: string;
    linkedin: string;
  };
  biography: string;
  image: Asset;
}

export interface Asset {
  url: string;
  title: string;
  id: string;
}

export interface Article {
  id: string;
  type: string;
  title: string;
  abstract: string;
  seo: {
    description: string;
    keywords: string;
  };
  slug: string;
  website: Website;
  published_at: string;
  authors: Author[];
  blocks: BlockDefinition[];
  images: {
    list: Asset;
    header: Asset;
  };
}

export interface ContentfulWebsite {
  name: string;
  domains: string[];
  path: string;
  culture: string;
  country_code: string;
  order: number;
  enabled: boolean;
}

export interface ContentfulAuthor {
  name: string;
  job_title: string;
  slug: string;
  social_twitter: string;
  social_facebook: string;
  social_linkedin: string;
  biography: string;
  image: ContentfulAsset;
}

export interface ContentfulPage {
  title: string;
  type: string;
  seo_description: string;
  seo_keywords: string;
  extends: string;
  path: string;
  website: ContentfulWebsite;
  layout: string;
  ttl: number;
  settings: object;
  published_at: string;
  blocks: Entry<any>[];
  code: string;
}

export interface BaseContentfulBlock {
  internal_title: string;
  container: string;
  order: number;
}

export interface ContentfulBlockText extends BaseContentfulBlock {
  title: string;
  subtitle: string;
  contents: string;
  mode: string;
  image: ContentfulAsset;
  image_position: string;
}

export interface ContentfulArticle {
  type: string;
  title: string;
  abstract: string;
  seo_description: string;
  seo_keywords: string;
  slug: string;
  website: Entry<ContentfulWebsite>;
  published_at: string;
  authors: Entry<ContentfulAuthor>[];
  blocks: Entry<any>[];
  image_list: ContentfulAsset;
  image_header: ContentfulAsset;
}
