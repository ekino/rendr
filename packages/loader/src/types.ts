import { Page, RendrCtx, PageType } from "@ekino/rendr-core";
import { AxiosRequestConfig } from "axios";

export type MaybePage = Promise<PageType> | PageType;

export type Loader = (
  ctx: RendrCtx,
  page: PageType,
  next?: (page?: MaybePage) => MaybePage
) => MaybePage;

export type PageBuilder = (ctx: RendrCtx, page: Page) => MaybePage;

export type RouteConfiguration = {
  path: string;
  matcher: (path: string) => {} | undefined;
  pageBuilder: PageBuilder | Loader;
};

export interface InMemorySettings {
  [index: string]: PageBuilder | Loader;
}

export type AxiosOptionsBuilder = (
  url: string,
  options: AxiosRequestConfig
) => { url: string; options: AxiosRequestConfig };
