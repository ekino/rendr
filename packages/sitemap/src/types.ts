export type Options = {
  readonly basePathIndex: string;
};
