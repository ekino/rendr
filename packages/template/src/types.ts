import React from "react";
import { BlockDefinition, Settings, Page, RequestCtx } from "@ekino/rendr-core";

export type ReactElement<P> =
  | React.ComponentClass<P>
  | React.StatelessComponent<P>;

// -- Template definition
export interface TemplateProps {
  containerRenderer: ContainerRenderer;
  page: Page;
  blocks: BlockDefinition[];
}

export type TemplateRegistry = (code: string) => ReactElement<TemplateProps>;

// -- Block definition
export interface BlockResult {
  component: ReactElement<any>;
  settings: Settings;
}

export type BlockRegistry = (code: string, setting: Settings) => BlockResult;

// -- Rendering helpers
export type ContainerRenderer = (
  area: string,
  blocks: BlockDefinition[]
) => JSX.Element[] | JSX.Element;

export type BlockRenderer = (
  block: BlockDefinition,
  key: string
) => JSX.Element;

export interface ComponentList {
  [index: string]: ReactElement<any>;
}
