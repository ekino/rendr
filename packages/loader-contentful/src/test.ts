import { Entry, ContentfulClientApi } from "contentful";
import { readFileSync } from "fs";

export function createDummyEntry<T>(data: T): Entry<T> {
  return {
    sys: {
      type: "Entry",
      id: "",
      createdAt: "",
      updatedAt: "",
      locale: "fr_FR",
      contentType: {
        sys: {
          type: "Link",
          linkType: "ContentType",
          id: "DummyEntry",
        },
      },
      space: {
        sys: {
          type: "Link",
          linkType: "Space",
          id: "DummySpace",
        },
      },
    },
    fields: data,
    toPlainObject: () => ({}),
    update: () => Promise.resolve(this),
  };
}

export function loadJson(path: string) {
  return JSON.parse(readFileSync(path, { encoding: "utf8" }));
}

export function getMockClient(
  methods: Partial<ContentfulClientApi>
): ContentfulClientApi {
  return {
    getAsset: async (id: string, query?: any) => {
      throw new Error("The method should not be called");
    },
    getAssets: async (query?: any) => {
      throw new Error("The method should not be called");
    },
    getContentType: async (id: string) => {
      throw new Error("The method should not be called");
    },
    getContentTypes: async (query?: any) => {
      throw new Error("The method should not be called");
    },
    getEntries: async <T>(query?: any) => {
      throw new Error("The method should not be called");
    },
    getEntry: async <T>(id: string, query?: any) => {
      throw new Error("The method should not be called");
    },
    getSpace: async () => {
      throw new Error("The method should not be called");
    },
    getLocales: async () => {
      throw new Error("The method should not be called");
    },
    parseEntries: async <T>(raw: any) => {
      throw new Error("The method should not be called");
    },
    sync: async (query: any) => {
      throw new Error("The method should not be called");
    },
    ...methods,
  };
}
