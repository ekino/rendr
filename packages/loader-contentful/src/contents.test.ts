import { loadJson, getMockClient } from "./test";
import { Website } from "./types";
import {
  GetWebsites,
  GetWebsite,
  GetPage,
  GetAuthor,
  GetArticle
} from "./contents";

describe("Contents loading from ", () => {
  it("test order function", () => {
    const values = [2, 2, 3, 1];

    const sorted = values.sort((a, b) => {
      return a === b ? 0 : a > b ? 1 : -1;
    });

    expect(sorted).toEqual([1, 2, 2, 3]);
  });

  it("Should load Websites", async () => {
    const client = getMockClient({
      getEntries: async <T>(query?: any) => {
        expect(query).toEqual({
          content_type: "rendr_website",
          limit: 100,
          skip: 0
        });

        return loadJson(`${__dirname}/__fixtures__/contents/websites.json`);
      }
    });

    expect(await GetWebsites(client)).toMatchSnapshot();
  });

  it("Should return existing website ", async () => {
    const client = getMockClient({
      getEntries: async <T>(query?: any) => {
        expect(query).toEqual({
          content_type: "rendr_website",
          limit: 100,
          skip: 0
        });

        return loadJson(`${__dirname}/__fixtures__/contents/websites.json`);
      }
    });

    expect(await GetWebsite(client, "ekino.com")).toMatchSnapshot();
  });

  it("Should return non existing website ", async () => {
    const client = getMockClient({
      getEntries: async <T>(query?: any) => {
        expect(query).toEqual({
          content_type: "rendr_website",
          limit: 100,
          skip: 0
        });

        return loadJson(`${__dirname}/__fixtures__/contents/websites.json`);
      }
    });

    try {
      await GetWebsite(client, "foobar.com");
      expect(true).toBeFalsy();
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
  });

  it("Should raise an exception as no page will be found", async () => {
    const client = getMockClient({
      getEntries: async <T>(query?: any) => {
        expect(query).toEqual({
          "fields.path": "/",
          "fields.website.sys.id": "632kl7enPots4PISSnD6DV",
          limit: 1,
          content_type: "rendr_page"
        });

        return loadJson(
          `${__dirname}/__fixtures__/contents/empty_collection.json`
        );
      }
    });

    const website: Website = {
      id: "632kl7enPots4PISSnD6DV",
      name: "ekino",
      domains: ["ekino.com"],
      path: "/",
      culture: "en_GB",
      countryCode: "GDB",
      order: 0,
      enabled: true
    };

    try {
      await GetPage(client, website, "/");
      expect(true).toBeFalsy();
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
  });

  it("Should return existing page ", async () => {
    const client = getMockClient({
      getEntries: async <T>(query?: any) => {
        expect(query).toEqual({
          "fields.path": "/",
          "fields.website.sys.id": "632kl7enPots4PISSnD6DV",
          limit: 1,
          content_type: "rendr_page"
        });

        return loadJson(`${__dirname}/__fixtures__/contents/pages.json`);
      }
    });

    const website: Website = {
      id: "632kl7enPots4PISSnD6DV",
      name: "ekino",
      domains: ["ekino.com"],
      path: "/",
      culture: "en_GB",
      countryCode: "GDB",
      order: 0,
      enabled: true
    };

    expect(await GetPage(client, website, "/")).toMatchSnapshot();
  });

  it("should return an Author", async () => {
    const client = getMockClient({
      getEntries: async <T>(query?: any) => {
        expect(query).toEqual({
          content_type: "rendr_author",
          "fields.slug": "thomas-rabaix",
          limit: 2
        });

        return loadJson(`${__dirname}/__fixtures__/contents/authors.json`);
      }
    });

    expect(await GetAuthor(client, "thomas-rabaix")).toMatchSnapshot();
  });

  it("should return an error with Author (duplicate slug)", async () => {
    const client = getMockClient({
      getEntries: async <T>(query?: any) => {
        expect(query).toEqual({
          content_type: "rendr_author",
          "fields.slug": "thomas-rabaix",
          limit: 2
        });

        return loadJson(
          `${__dirname}/__fixtures__/contents/duplicate_authors.json`
        );
      }
    });

    try {
      await GetAuthor(client, "thomas-rabaix");
      expect(true).toBeFalsy();
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
  });

  it("should return an article", async () => {
    let call = 0;
    const client = getMockClient({
      getEntries: async <T>(query?: any) => {
        call++;

        if (call === 1) {
          expect(query).toEqual({
            content_type: "rendr_website",
            limit: 100,
            skip: 0
          });

          return loadJson(`${__dirname}/__fixtures__/contents/websites.json`);
        }

        if (call === 2) {
          query["fields.published_at[lte]"] = "fail to mock the date...";

          expect(query).toEqual({
            content_type: "rendr_article",
            "fields.published_at[lte]": "fail to mock the date...",
            "fields.website.sys.id": "632kl7enPots4PISSnD6DV",
            "fields.slug": "headless-cms-challenge",
            limit: 2
          });

          return loadJson(`${__dirname}/__fixtures__/contents/article.json`);
        }
      }
    });

    expect(
      await GetArticle(client, "headless-cms-challenge", {
        domain: "ekino.co.uk"
      })
    ).toMatchSnapshot();
  });
});
