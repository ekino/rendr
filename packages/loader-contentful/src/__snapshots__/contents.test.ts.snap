// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contents loading from  Should load Websites 1`] = `
Array [
  Object {
    "fields": Object {
      "country_code": "FRA",
      "culture": "fr_FR",
      "domains": Array [
        "ekino.com",
      ],
      "enabled": true,
      "name": "Ekino",
      "order": 1,
      "path": "/",
    },
    "sys": Object {
      "contentType": Object {
        "sys": Object {
          "id": "rendr_website",
          "linkType": "ContentType",
          "type": "Link",
        },
      },
      "createdAt": "2019-08-21T11:00:42.276Z",
      "environment": Object {
        "sys": Object {
          "id": "master",
          "linkType": "Environment",
          "type": "Link",
        },
      },
      "id": "2g5lXpzZ0jPPwQViziSjtK",
      "locale": "en-US",
      "revision": 1,
      "space": Object {
        "sys": Object {
          "id": "f5bx295b472q",
          "linkType": "Space",
          "type": "Link",
        },
      },
      "type": "Entry",
      "updatedAt": "2019-08-21T11:00:42.276Z",
    },
  },
  Object {
    "fields": Object {
      "country_code": "GDB",
      "culture": "en_GB",
      "enabled": true,
      "name": "Ekino UK",
      "order": 2,
      "path": "/uk",
    },
    "sys": Object {
      "contentType": Object {
        "sys": Object {
          "id": "rendr_website",
          "linkType": "ContentType",
          "type": "Link",
        },
      },
      "createdAt": "2019-08-21T12:43:36.952Z",
      "environment": Object {
        "sys": Object {
          "id": "master",
          "linkType": "Environment",
          "type": "Link",
        },
      },
      "id": "632kl7enPots4PISSnD6DV",
      "locale": "en-US",
      "revision": 1,
      "space": Object {
        "sys": Object {
          "id": "f5bx295b472q",
          "linkType": "Space",
          "type": "Link",
        },
      },
      "type": "Entry",
      "updatedAt": "2019-08-21T12:43:36.952Z",
    },
  },
]
`;

exports[`Contents loading from  Should raise an exception as no page will be found 1`] = `[Error: Not Found Exception]`;

exports[`Contents loading from  Should return existing page  1`] = `
Object {
  "fields": Object {
    "path": "/",
    "seo_description": "/",
    "seo_keywords": "/",
    "title": "homepage",
    "website": Object {
      "fields": Object {
        "country_code": "GDB",
        "culture": "en_GB",
        "enabled": true,
        "name": "Ekino UK",
        "path": "/uk",
      },
      "sys": Object {
        "contentType": Object {
          "sys": Object {
            "id": "rendr_website",
            "linkType": "ContentType",
            "type": "Link",
          },
        },
        "createdAt": "2019-08-21T12:43:36.952Z",
        "environment": Object {
          "sys": Object {
            "id": "master",
            "linkType": "Environment",
            "type": "Link",
          },
        },
        "id": "632kl7enPots4PISSnD6DV",
        "locale": "en-US",
        "revision": 1,
        "space": Object {
          "sys": Object {
            "id": "f5bx295b472q",
            "linkType": "Space",
            "type": "Link",
          },
        },
        "type": "Entry",
        "updatedAt": "2019-08-21T12:43:36.952Z",
      },
    },
  },
  "sys": Object {
    "contentType": Object {
      "sys": Object {
        "id": "rendr_page",
        "linkType": "ContentType",
        "type": "Link",
      },
    },
    "createdAt": "2019-08-21T15:43:17.711Z",
    "environment": Object {
      "sys": Object {
        "id": "master",
        "linkType": "Environment",
        "type": "Link",
      },
    },
    "id": "28dAkVDQuMmGBAX7IEOJDd",
    "locale": "en-US",
    "revision": 2,
    "space": Object {
      "sys": Object {
        "id": "f5bx295b472q",
        "linkType": "Space",
        "type": "Link",
      },
    },
    "type": "Entry",
    "updatedAt": "2019-08-21T15:43:44.995Z",
  },
}
`;

exports[`Contents loading from  Should return existing website  1`] = `
Object {
  "fields": Object {
    "country_code": "FRA",
    "culture": "fr_FR",
    "domains": Array [
      "ekino.com",
    ],
    "enabled": true,
    "name": "Ekino",
    "order": 1,
    "path": "/",
  },
  "sys": Object {
    "contentType": Object {
      "sys": Object {
        "id": "rendr_website",
        "linkType": "ContentType",
        "type": "Link",
      },
    },
    "createdAt": "2019-08-21T11:00:42.276Z",
    "environment": Object {
      "sys": Object {
        "id": "master",
        "linkType": "Environment",
        "type": "Link",
      },
    },
    "id": "2g5lXpzZ0jPPwQViziSjtK",
    "locale": "en-US",
    "revision": 1,
    "space": Object {
      "sys": Object {
        "id": "f5bx295b472q",
        "linkType": "Space",
        "type": "Link",
      },
    },
    "type": "Entry",
    "updatedAt": "2019-08-21T11:00:42.276Z",
  },
}
`;

exports[`Contents loading from  Should return non existing website  1`] = `[Error: Not Found Exception]`;

exports[`Contents loading from  should return an Author 1`] = `
Object {
  "fields": Object {
    "biography": "Just a guy...",
    "image": Object {
      "fields": Object {
        "file": Object {
          "contentType": "image/jpeg",
          "details": Object {
            "image": Object {
              "height": 1023,
              "width": 969,
            },
            "size": 247335,
          },
          "fileName": "DSCF2267.jpg",
          "url": "//images.ctfassets.net/f5bx295b472q/2pSrRlUsweaxu4tW7w1uVv/dd3e01d59c68d8b07101ddf953a120f7/DSCF2267.jpg",
        },
        "title": "Thomas Rabaix's Picture",
      },
      "sys": Object {
        "createdAt": "2019-08-21T10:59:57.177Z",
        "environment": Object {
          "sys": Object {
            "id": "master",
            "linkType": "Environment",
            "type": "Link",
          },
        },
        "id": "2pSrRlUsweaxu4tW7w1uVv",
        "locale": "en-US",
        "revision": 1,
        "space": Object {
          "sys": Object {
            "id": "f5bx295b472q",
            "linkType": "Space",
            "type": "Link",
          },
        },
        "type": "Asset",
        "updatedAt": "2019-08-21T10:59:57.177Z",
      },
    },
    "job_title": "CTO London",
    "name": "Thomas Rabaix",
    "slug": "thomas-rabaix",
    "social_twitter": "th0masr",
  },
  "sys": Object {
    "contentType": Object {
      "sys": Object {
        "id": "rendr_author",
        "linkType": "ContentType",
        "type": "Link",
      },
    },
    "createdAt": "2019-08-21T11:00:16.201Z",
    "environment": Object {
      "sys": Object {
        "id": "master",
        "linkType": "Environment",
        "type": "Link",
      },
    },
    "id": "4JSAPT1jf0J8XJa0cg1aCa",
    "locale": "en-US",
    "revision": 1,
    "space": Object {
      "sys": Object {
        "id": "f5bx295b472q",
        "linkType": "Space",
        "type": "Link",
      },
    },
    "type": "Entry",
    "updatedAt": "2019-08-21T11:00:16.201Z",
  },
}
`;

exports[`Contents loading from  should return an article 1`] = `
Object {
  "fields": Object {
    "abstract": "The challenge of using a Headless CMS but still want to control how the template is managed",
    "authors": Array [
      Object {
        "fields": Object {
          "biography": "Just a guy...",
          "image": Object {
            "fields": Object {
              "file": Object {
                "contentType": "image/jpeg",
                "details": Object {
                  "image": Object {
                    "height": 1023,
                    "width": 969,
                  },
                  "size": 247335,
                },
                "fileName": "DSCF2267.jpg",
                "url": "//images.ctfassets.net/f5bx295b472q/2pSrRlUsweaxu4tW7w1uVv/dd3e01d59c68d8b07101ddf953a120f7/DSCF2267.jpg",
              },
              "title": "Thomas Rabaix's Picture",
            },
            "sys": Object {
              "createdAt": "2019-08-21T10:59:57.177Z",
              "environment": Object {
                "sys": Object {
                  "id": "master",
                  "linkType": "Environment",
                  "type": "Link",
                },
              },
              "id": "2pSrRlUsweaxu4tW7w1uVv",
              "locale": "en-US",
              "revision": 1,
              "space": Object {
                "sys": Object {
                  "id": "f5bx295b472q",
                  "linkType": "Space",
                  "type": "Link",
                },
              },
              "type": "Asset",
              "updatedAt": "2019-08-21T10:59:57.177Z",
            },
          },
          "job_title": "CTO London",
          "name": "Thomas Rabaix",
          "slug": "thomas-rabaix",
          "social_twitter": "th0masr",
        },
        "sys": Object {
          "contentType": Object {
            "sys": Object {
              "id": "rendr_author",
              "linkType": "ContentType",
              "type": "Link",
            },
          },
          "createdAt": "2019-08-21T11:00:16.201Z",
          "environment": Object {
            "sys": Object {
              "id": "master",
              "linkType": "Environment",
              "type": "Link",
            },
          },
          "id": "4JSAPT1jf0J8XJa0cg1aCa",
          "locale": "en-US",
          "revision": 1,
          "space": Object {
            "sys": Object {
              "id": "f5bx295b472q",
              "linkType": "Space",
              "type": "Link",
            },
          },
          "type": "Entry",
          "updatedAt": "2019-08-21T11:00:16.201Z",
        },
      },
    ],
    "blocks": Array [
      Object {
        "fields": Object {
          "container": "header",
          "contents": "The article contents with the markdown format! ",
          "internal_title": "The main contents",
        },
        "sys": Object {
          "contentType": Object {
            "sys": Object {
              "id": "rendr_block_text",
              "linkType": "ContentType",
              "type": "Link",
            },
          },
          "createdAt": "2019-08-21T10:58:31.192Z",
          "environment": Object {
            "sys": Object {
              "id": "master",
              "linkType": "Environment",
              "type": "Link",
            },
          },
          "id": "57UPFciwX1EEPgkZNh0gXn",
          "locale": "en-US",
          "revision": 1,
          "space": Object {
            "sys": Object {
              "id": "f5bx295b472q",
              "linkType": "Space",
              "type": "Link",
            },
          },
          "type": "Entry",
          "updatedAt": "2019-08-21T10:58:31.192Z",
        },
      },
      Object {
        "fields": Object {
          "container": "header",
          "contents": "The second part of the journey",
          "internal_title": "The second block",
        },
        "sys": Object {
          "contentType": Object {
            "sys": Object {
              "id": "rendr_block_text",
              "linkType": "ContentType",
              "type": "Link",
            },
          },
          "createdAt": "2019-08-21T11:04:53.316Z",
          "environment": Object {
            "sys": Object {
              "id": "master",
              "linkType": "Environment",
              "type": "Link",
            },
          },
          "id": "3CxKOYXVpLCPCwNCXGdn4N",
          "locale": "en-US",
          "revision": 1,
          "space": Object {
            "sys": Object {
              "id": "f5bx295b472q",
              "linkType": "Space",
              "type": "Link",
            },
          },
          "type": "Entry",
          "updatedAt": "2019-08-21T11:04:53.316Z",
        },
      },
    ],
    "image_header": Object {
      "fields": Object {
        "file": Object {
          "contentType": "image/png",
          "details": Object {
            "image": Object {
              "height": 782,
              "width": 2646,
            },
            "size": 1426695,
          },
          "fileName": "Screenshot 2019-08-21 at 12.02.06.png",
          "url": "//images.ctfassets.net/f5bx295b472q/5vV7exdcps9vwi2I1XEFfP/da39b569566d118b2711f8e5ec210c76/Screenshot_2019-08-21_at_12.02.06.png",
        },
        "title": "Header image",
      },
      "sys": Object {
        "createdAt": "2019-08-21T11:03:21.588Z",
        "environment": Object {
          "sys": Object {
            "id": "master",
            "linkType": "Environment",
            "type": "Link",
          },
        },
        "id": "5vV7exdcps9vwi2I1XEFfP",
        "locale": "en-US",
        "revision": 1,
        "space": Object {
          "sys": Object {
            "id": "f5bx295b472q",
            "linkType": "Space",
            "type": "Link",
          },
        },
        "type": "Asset",
        "updatedAt": "2019-08-21T11:03:21.588Z",
      },
    },
    "image_list": Object {
      "fields": Object {
        "file": Object {
          "contentType": "image/png",
          "details": Object {
            "image": Object {
              "height": 624,
              "width": 1198,
            },
            "size": 463423,
          },
          "fileName": "Screenshot 2019-08-21 at 12.02.00.png",
          "url": "//images.ctfassets.net/f5bx295b472q/4BRlYkVTQwQBcGb2Gft8Nb/700ef1992bd71efbe6ce42f6741a1ca2/Screenshot_2019-08-21_at_12.02.00.png",
        },
        "title": "Main image",
      },
      "sys": Object {
        "createdAt": "2019-08-21T11:02:44.877Z",
        "environment": Object {
          "sys": Object {
            "id": "master",
            "linkType": "Environment",
            "type": "Link",
          },
        },
        "id": "4BRlYkVTQwQBcGb2Gft8Nb",
        "locale": "en-US",
        "revision": 2,
        "space": Object {
          "sys": Object {
            "id": "f5bx295b472q",
            "linkType": "Space",
            "type": "Link",
          },
        },
        "type": "Asset",
        "updatedAt": "2019-08-21T11:04:02.200Z",
      },
    },
    "published_at": "2019-08-15T00:00",
    "seo_description": "The description",
    "seo_keywords": "The keywords",
    "slug": "headless-cms-challenge",
    "title": "Headless CMS challenge",
    "website": Object {
      "fields": Object {
        "country_code": "FRA",
        "culture": "fr_FR",
        "domains": Array [
          "ekino.com",
        ],
        "enabled": true,
        "name": "Ekino",
        "path": "/",
      },
      "sys": Object {
        "contentType": Object {
          "sys": Object {
            "id": "rendr_website",
            "linkType": "ContentType",
            "type": "Link",
          },
        },
        "createdAt": "2019-08-21T11:00:42.276Z",
        "environment": Object {
          "sys": Object {
            "id": "master",
            "linkType": "Environment",
            "type": "Link",
          },
        },
        "id": "2g5lXpzZ0jPPwQViziSjtK",
        "locale": "en-US",
        "revision": 1,
        "space": Object {
          "sys": Object {
            "id": "f5bx295b472q",
            "linkType": "Space",
            "type": "Link",
          },
        },
        "type": "Entry",
        "updatedAt": "2019-08-21T11:00:42.276Z",
      },
    },
  },
  "sys": Object {
    "contentType": Object {
      "sys": Object {
        "id": "rendr_article",
        "linkType": "ContentType",
        "type": "Link",
      },
    },
    "createdAt": "2019-08-21T11:04:56.829Z",
    "environment": Object {
      "sys": Object {
        "id": "master",
        "linkType": "Environment",
        "type": "Link",
      },
    },
    "id": "4yrf1lgxv3VY8K1jpjkog0",
    "locale": "en-US",
    "revision": 1,
    "space": Object {
      "sys": Object {
        "id": "f5bx295b472q",
        "linkType": "Space",
        "type": "Link",
      },
    },
    "type": "Entry",
    "updatedAt": "2019-08-21T11:04:56.829Z",
  },
}
`;

exports[`Contents loading from  should return an error with Author (duplicate slug) 1`] = `[Error: Not Found Exception]`;
